### PR TITLE
Review download code

### LINKED ISSUE #1399: ngclient: review download code
(this is about code currently in experimental-client branch, hopefully soon in develop)

After we removed mirrors support, we should look through the download code 
* Add type annotations
* replace old style docstrings
* see if code can be simplified
* consider removing average download speed check (it is useless in practice, will trigger false positives if the limit was usefully low and makes the code complex). It also prevents setting chunk size to a reasonable value (it's currently way too large but I'm afraid to lower it because the average speed check then becomes even more volatile for the first chunks -- if the limit is even close to reasonable, it could trigger for the first chunk much easier than later chunks just because that's how statistics works)
* consider getting rid of strict_required_length: I do not see the value, and it makes the code more complex (if on the other hand there is actual value in length checks, we should implement them in MetadataBundle: the spec does not require them but they would be easy to add)

### PR DESCRIPTION
**Pull request by sechkova**
Fixes #1399

**Description of the changes being introduced by the pull request**:

_This PR builds on top of #1408 and will remain a draft until its merge._

- simplify download code, remove average download speed check
- wrap the download functionality in `FileDownload` class
- simplify `RequestsFetcher.fetch()`
- add type annotations

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by jku** [comment]
Looks great at first glance. 

This maybe a dumb question but is there something that prevents us from putting the new FileDownloader functionality directly into FetcherInterface?

Are the RequestsFetcher improvements dependent on the other changes or just included in same PR for convenience?

**Comment by sechkova** [comment]
> This maybe a dumb question but is there something that prevents us from putting the new FileDownloader functionality directly into FetcherInterface?
> 
Only me not thinking enough about it ... :roll_eyes: 

> Are the RequestsFetcher improvements dependent on the other changes or just included in same PR for convenience?

No, they are not dependent, I just consider fetcher as part of the download code.

I do like the idea of simplifying further and keeping only a single download/fetcher module instead of the two. How should we do it? Build on top of this PR, a separate PR or I can rewrite these commits here and directly move download.py functions into FetcherInterface?

**Comment by jku** [comment]
> I do like the idea of simplifying further and keeping only a single download/fetcher module instead of the two. How should we do it? Build on top of this PR, a separate PR or I can rewrite these commits here and directly move download.py functions into FetcherInterface?

I think merging a PR that adds a new object would be weird if we intend to remove it soon in a followup PR... otherwise whatever works for you I think

**Comment by sechkova** [comment]
Rewrote part of the commits here. Now the simplified download functions are made part of `FetcherInterface` and the `download.py` module is removed.

**Comment by sechkova** [comment]
I don't know what is blocking GitHub actions here?

**Comment by jku** [comment]
> I don't know what is blocking GitHub actions here?

quite strange. It never actually started the action (https://github.com/theupdateframework/tuf/actions) so I can't even restart it.

Looks really nice btw... Net result of removing 200 lines and ending up with a cleaner implementation is not too bad.

File: tuf/ngclient/fetcher.py [hunk 1]
```diff
@@ -40,3 +48,51 @@ def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
...
+        Yields:
+          A file object that points to the contents of 'url'.
+        """
+        # 'url.replace('\\', '/')' is needed for compatibility with
+        # Windows-based systems, because they might use back-slashes in place
+        # of forward-slashes. This converts it to the common format.
+        # unquote() replaces %xx escapes in a url with their single-character
+        # equivalent. A back-slash may beencoded as %5c in the url, which
+        # should also be replaced with a forward slash.
+        url = parse.unquote(url).replace("\\", "/")
```
Comment (jku) [inline tuf/ngclient/fetcher.py:74]: let's get rid of this: this component should expect to get a real URL and should not do this.

Do we need a new issue to handle possible issues in higher level code?

File: tuf/ngclient/fetcher.py [hunk 2]
```diff
@@ -40,3 +48,51 @@ def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
             A bytes iterator
         """
         raise NotImplementedError  # pragma: no cover
+
+    @contextmanager
+    def download_file(self, url: str, required_length: int) -> Iterator[IO]:
```
Comment (jku) [inline tuf/ngclient/fetcher.py]: required_length is actually more like max_length, right?

File: tuf/ngclient/fetcher.py [hunk 3]
```diff
@@ -40,3 +48,51 @@ def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
...
+          url: a URL string that represents the location of the file.
+          required_length: an integer value representing the length of
+              the file or an upper boundary.
+
+        Raises:
+          DownloadLengthMismatchError: a mismatch of observed vs expected
+              lengths while downloading the file.
+
+        Yields:
+          A file object that points to the contents of 'url'.
```
Comment (jku) [inline tuf/ngclient/fetcher.py]: I wonder if it's worth mentioning that it's a TemporaryFile?

File: tuf/ngclient/fetcher.py [hunk 4]
```diff
@@ -40,3 +48,51 @@ def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
...
+                temp_file.write(chunk)
+                number_of_bytes_received += len(chunk)
+            if number_of_bytes_received > required_length:
+                raise exceptions.DownloadLengthMismatchError(
+                    required_length, number_of_bytes_received
+                )
+            temp_file.seek(0)
+            yield temp_file
+
+    def download_bytes(self, url: str, required_length: int) -> bytes:
```
Comment (jku) [inline tuf/ngclient/fetcher.py]: (same comment on required_length naming)

File: tuf/ngclient/fetcher.py [hunk 5]
```diff
@@ -39,3 +48,51 @@ def fetch(self, url, required_length):
...
+        up to 'required_length'.
+
+        Args:
+          url: a URL string that represents the location of the file.
+          required_length: an integer value representing the length of
+              the file or an upper boundary.
+
+        Raises:
+          DownloadLengthMismatchError: a mismatch of observed vs expected
+              lengths while downloading the file.
```
Comment (jku) [inline tuf/ngclient/fetcher.py]:
```suggestion
          DownloadLengthMismatchError: File length was larger than <name of argument>
```

File: tuf/ngclient/_internal/requests_fetcher.py [hunk 6]
```diff
@@ -86,58 +87,62 @@ def fetch(self, url, required_length):
...
+                # large file in one shot. Before beginning the round, sleep
+                # (if set) for a short amount of time so that the CPU is not
+                # hogged in the while loop.
+                if settings.SLEEP_BEFORE_ROUND:
+                    time.sleep(settings.SLEEP_BEFORE_ROUND)
+
+                read_amount = min(
+                    settings.CHUNK_SIZE,
+                    required_length - bytes_received,
+                )
```
Comment (jku) [inline tuf/ngclient/_internal/requests_fetcher.py]: let's get rid of settings: 
* CHUNK_SIZE: this is a RequestsFetcher internal detail. make it an instance attribute (maybe with a smaller default than the current huge number)?
* SLEEP_BEFORE_ROUND: if this truly is an issue we should fix the issue -- there is no reason for us to be running a hot while loop for downloading things, that's just crazy.  #338 documents that that is indeed how the code used to work... but that was a long time ago, and the code looked entirely different. I'd say let's spend a bit of time making sure the loop is sane (and doesn't hog cpu) and let's remove SLEEP_BEFORE_ROUND

**Review by jku — COMMENTED** [review]
I like it a lot: this makes download code much easier to understand (and I agree with the decisions made)! Left several comments but the only one that I think needs to be fixed here is the argument name: _required_length_ does not seem to represent what it is.

Feel free to file issues for the other things

File: tuf/ngclient/_internal/requests_fetcher.py (hunk 6 above)
Comment (jku) [reply tuf/ngclient/_internal/requests_fetcher.py]: for reference, pip and requests use default chunk size of 1024*10

File: tuf/ngclient/fetcher.py (hunk 2 above)
Comment (sechkova) [reply tuf/ngclient/fetcher.py]: yes, it is the upper boundary.

File: tuf/ngclient/_internal/requests_fetcher.py (hunk 6 above)
Comment (jku) [reply tuf/ngclient/_internal/requests_fetcher.py]: this is alread being handled 1470 so maybe forget the comment here -- we should get rid of sleep before round but doesn't need to happen here.

File: tuf/ngclient/fetcher.py (hunk 1 above)
Comment (sechkova) [reply tuf/ngclient/fetcher.py:74]: Opened a new issue although there are several ones related to paths and urls in Updater already: https://github.com/theupdateframework/tuf/issues/1483

File: tuf/ngclient/_internal/requests_fetcher.py (hunk 6 above)
Comment (sechkova) [reply tuf/ngclient/_internal/requests_fetcher.py]: I opened https://github.com/theupdateframework/tuf/issues/1484 for it and we can continue discussing settings in the #1470 thread.

File: tuf/ngclient/fetcher.py (hunk 1 above)
Comment (jku) [reply tuf/ngclient/fetcher.py:74]: oh you might be right about multiple bugs -- I probably file one every time I see this code

**Comment by sechkova** [comment]
I think all suggestions are added with 959ab5a.

**Comment by sechkova** [comment]
Rebased on latest changes in develop.

File: tuf/ngclient/_internal/requests_fetcher.py [hunk 7]
```diff
@@ -53,7 +53,7 @@ def __init__(self):
         self.chunk_size: int = 400000  # bytes
         self.sleep_before_round: Optional[int] = None
 
-    def fetch(self, url, required_length):
+    def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
```
Comment (jku) [inline tuf/ngclient/_internal/requests_fetcher.py]: I guess this is more `max_length` than `required_length` as well?

the docstring could be better as well

File: tuf/ngclient/fetcher.py [hunk 8]
```diff
@@ -20,7 +29,7 @@ class FetcherInterface:
     __metaclass__ = abc.ABCMeta
 
     @abc.abstractmethod
-    def fetch(self, url, required_length):
+    def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
```
Comment (jku) [inline tuf/ngclient/fetcher.py]: maybe should be max_length (also docstring could be better)?

This is sort of API already but the change maybe worth it as required_length really is misleading...

**Review by jku — APPROVED** [review]
Left comments/question about naming of the argument to Fetcher.fetch() as well... but I'm approving as is, this is a very good improvement on download code IMO

**Comment by sechkova** [comment]
Amended the last commit 0eff004 with the suggestion.

**Review by MVrachev — APPROVED** [review]
Great job!
I liked all of the simplifications.
It's easier to read now.
It looks good to me as well.

**Pull request merged**